Crypto/PublicKey+Apple.mm

```diff
@@ -262,58 +262,75 @@ virtual int _sign(int/*mbedtls_md_type_t*/ mbedDigestAlgorithm,
-        checkOSStatus(SecCertificateCopyPublicKey(certRef, &publicKeyRef),
-                      "SecCertificateCopyPublicKey", "get private key from keychain");
-
-        SecIdentityRef identityRef;
-        checkOSStatus(SecIdentityCreateWithCertificate(nullptr, certRef, &identityRef),
-                      "SecIdentityCreateWithCertificate", "get private key from keychain");
-        CFAutorelease(identityRef);
-        SecKeyRef privateKeyRef;
-        checkOSStatus(SecIdentityCopyPrivateKey(identityRef, &privateKeyRef),
-                          "SecIdentityCopyPrivateKey", "get private key from keychain");
-        auto keySize = unsigned(8 * SecKeyGetBlockSize(privateKeyRef));
-        return new KeychainKeyPair(keySize, publicKeyRef, privateKeyRef);
+        @autoreleasepool {
+            SecCertificateRef certRef = SecCertificateCreateWithData(kCFAllocatorDefault,
+                                                                     (CFDataRef)cert->data().copiedNSData());
+            if (!certRef)
+                throwMbedTLSError(MBEDTLS_ERR_X509_INVALID_FORMAT); // impossible?
+            CFAutorelease(certRef);
+            
+            // Get public key from the certificate:
+            SecKeyRef publicKeyRef;
+            if (@available(iOS 12, macOS 10.14, *)) {
+                publicKeyRef = SecCertificateCopyKey(certRef);
+            } else {
+#if TARGET_OS_IOS
+                if (@available(iOS 10.3, *))
+                    publicKeyRef = SecCertificateCopyPublicKey(certRef);
+                else
+                    throwMbedTLSError(MBEDTLS_ERR_X509_FEATURE_UNAVAILABLE);
+#else
+                checkOSStatus(SecCertificateCopyPublicKey(certRef, &publicKeyRef),
+                              "SecCertificateCopyPublicKey", "get private key from keychain");
+#endif
+            }
+            
+            if (!publicKeyRef)
+                return nullptr;
+            
+            // Get public key hash from kSecAttrApplicationLabel attribute:
+            // See: https://developer.apple.com/documentation/security/ksecattrapplicationlabel
+            NSDictionary* attrs = CFBridgingRelease(SecKeyCopyAttributes(publicKeyRef));
+            NSData* publicKeyHash = [attrs objectForKey: (id)kSecAttrApplicationLabel];
+            if (!publicKeyHash)
+                throwMbedTLSError(MBEDTLS_ERR_X509_INVALID_FORMAT);
+            
+            // Lookup private key by using the public key hash:
+            auto privateKeyRef = (SecKeyRef)findInKeychain(@{
+                (id)kSecClass:                  (id)kSecClassKey,
+                (id)kSecAttrKeyType:            (id)kSecAttrKeyTypeRSA,
+                (id)kSecAttrKeyClass:           (id)kSecAttrKeyClassPrivate,
+                (id)kSecAttrApplicationLabel:   publicKeyHash,
+                (id)kSecReturnRef:              @YES
+            });
+            if (!privateKeyRef) {
+                CFAutorelease(publicKeyRef);
```

Comment:
why are we auto releasing here, instead of outside

Reply:
Because we want to release it here only when there is an error.

Reply:
It was supposed to be CFRelease even though CFAutoRelease when returning is fine too.